### PR TITLE
V2: Adds local notifications for geofence triggers and optimizes battery usage.

### PR DESCRIPTION
**Pull request by sophiabersani**
Stories:
https://www.pivotaltracker.com/story/show/124851181
https://www.pivotaltracker.com/story/show/125221331
https://www.pivotaltracker.com/story/show/125946737
https://www.pivotaltracker.com/story/show/125946995
- Adds Local push notifications for geofence crossings when the app is in the background.
  - Implements vibration support for local notifications. 
- Adds an energy-saving configuration to the geolocation service. Triggered whenever an app goes into the background.

Update:
- Fixes a bug with geofence triggers remaining enabled after moving out of range, in certain conditions.
- Fixes a bug with forms being loaded out of order when completing geofence Submissions.

### CODE REVIEW

File: js/api/BackgroundProcess.js [hunk 1]
```diff
@@ -11,55 +13,156 @@ export const BackgroundGeolocation = Platform.OS === 'ios' ?
...
     BackgroundGeolocation.on('error', function(error) {
-      // printTimelog('error')
+      printTimelog('error');
       console.log(error.type + " Error: " + error.code)
-    })
+    });
 
     // Create initial geofence hooks.
-    setupGeofences()
+    setupGeofences(()=>{
```
Comment (dan-nyanko) [inline js/api/BackgroundProcess.js:122]: I had several issues on the emulator and real devices if the geofences were added before the plugin 'starts'.  In fact, one would think the plugin is broken as the geofences would never trigger. Please see: https://github.com/ecohealthalliance/GoodQuestion/blob/geolocation-tests/App.js#L68-L77.

Comment (dan-nyanko) [inline js/api/BackgroundProcess.js:122]: I just traced the method setupGeofences and notices that its making a call to BackgroundGeolocation.start() then adding geofences.  This callback is also starting BackgroundGeolocation.  Please take a second look at these methods and fix.

File: js/api/BackgroundProcess.js [hunk 2]
```diff
@@ -11,55 +13,156 @@ export const BackgroundGeolocation = Platform.OS === 'ios' ?
...
+    });
   })
 }
 
+/**
+ * Configures the background service to use less energy when the app is not on the foreground.
+ * Ignores 'inactive' state. (transitioning, using phone, etc.)
+ * @param  {string} state The new state the app is transitioning into: 'active', 'inactive', 'background'
+ */
+export function handleAppStateChange(state) {
```
Comment (dan-nyanko) [inline js/api/BackgroundProcess.js:136]: Please explain why on L142 `BackgroundGeolocation.start((state) => {` and L155 is `BackgroundGeolocation.startGeofences((state) => {`.

**Comment by dan-nyanko** [comment]
Its really difficult to track the actual changes done in this PR due to the way branching/merging was done. Maybe this could be cleaned up by doing:

```
git checkout geofence-notifications
git branch -m geofence-notifications-old
git checkout master-v2
git checkout -b geofence-notifications
git merge --squash geofence-notifications-old
git commit
```

Then update to github with `git push --force origin geofence-notifications`

**Comment by sophiabersani** [comment]
@dan-nyanko Updated.

File: js/api/BackgroundProcess.js (hunk 1 above)
Comment (sophiabersani) [inline js/api/BackgroundProcess.js:122]: Calling a second `.start()` after setting up geofences was the only way I found to detect a geofence I was sitting on top of when accepting a survey, in this version of the script.

PR #141 upgrades React to 0.29 (changes Android templating) and uses the 2.0 version of this library.
Can you check if you have more success getting these to work on this new version? So far it seems a lot more stable on my end.

File: js/api/BackgroundProcess.js (hunk 2 above)
Comment (sophiabersani) [inline js/api/BackgroundProcess.js:136]: Battery optimization on Android devices, it has an option on their native API in which the GPS only keeps track of geofence crossings.

This along with the other config changes made my app go from 25-30% battery usage during field tests to around 3%.

iOS does not have this option and uses an API called "significant changes" instead.

Comment (dan-nyanko) [inline js/api/BackgroundProcess.js:136]: I mean why call `.start` vs. `.startGeofences`? Is `.startGeofences` even a method?

Comment (sophiabersani) [inline js/api/BackgroundProcess.js:136]: It's a method, documented on the android version of the library's API. (not available on iOS)
`.start` allows us to keep track of the user more accurately, for features such as sending their current location after they fill up a form. Switching to `.startGeofences` mode when the app goes on the background helps conserve battery as we do not need the other information at that stage.

**Pull request merged**